---
title: "Data Analysis Write Up"
author: "Peter Laurin"
date: "12/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
#library(reticulate) - python implementation - see below
library(ape)
library(MASS)
library(car)
#setwd('Needed_for_Analysis/')
#Sys.setenv(RETICULATE_PYTHON = "python/bin/python")
```
To begin with, a lot of larger data was used to generate the final analysis, so I can't 
bring you from my data all the way to my conclusions, but I'll do as much as I can using the
parsed/smaller data.

I also use a mix of python and R for this analysis. Because I run a lot of pretty heavy python scripts, I'm choosing to comment them out for the final version, (you won't have to download and use my python environment, which is several hundred MB).  


First, to conduct GWAS we need a genotype. This would be our phen_prelim_Oct2019.csv file, generated from infecting p25.C2 with all of the strains below. 
```{bash}
head -n 3 phen_prelim_Oct2019.csv
```
Blups (best linear unbiased prediction) are already regressed with certain covariates, including block, plate, day harvested, and other nusiance variables to get a phenotype for GWAS, as including several covariates in a GWAS implementation in pyseer can be tricky.


Let's write it in a pyseer-friendly way. 
```{bash}
cat phen_prelim_Oct2019.csv | cut -d ',' -f 2,3 | tr ',' '\t' > phen_prelim_Oct2019.tsv
head -n 3 phen_prelim_Oct2019.tsv

cat phen_prelim_Oct2019.tsv | cut  -f 1 | tail -n +2 > pheno_strains.txt
```

Phenotype done! I also quickly made a file including all the strains in our phenotype for further reference. For our genetic variants, we'll use presence/absence data for our orthology clusters generated by PanX (Ding, Baumdicker, & Neher, 2018). 

```{R}
#this block is a little intensive, so feel free to skip it!
fasta_data <- read.FASTA('genePresence.fasta') %>% as.character()
pheno_strains <- read_lines('pheno_strains.txt')

num_genes <- length(fasta_data$p2.E10)
fasta_data <- fasta_data[names(fasta_data) %in% pheno_strains]

pa_data <- tibble(.rows = num_genes)

for (strain in names(fasta_data)){
  for (sites in fasta_data[strain]){
    pa_data[strain] <- sites
    pa_data[strain][pa_data[strain] == 'c'] <- '1'
    pa_data[strain][pa_data[strain] == 't'] <-  '0'
    }
  }

pa_data <- mutate_all(pa_data, function(x) as.numeric(as.character(x)))
pa_data <- add_column(pa_data, Gene = as.character(1:num_genes), .before = 1)

write_tsv(pa_data, path = 'pa_data.tsv')
```

We have our genetic data, for all 72,397 clusters, data on whether our strains contain the variant or not.

Let's attempt a Kinship Matrix, 225 * 225 matrix to reduce population structure confounding in our GWAS. We'll implement this in pyseer, using -lmm. But first, we need a tree, ideally with only our phenotype strains, to keep our matrix as coherent as possible.
```{R}
full_tree <- read.tree(file = 'agloss_400.nwk')
good_tips <- full_tree$tip.label[full_tree$tip.label %in% pheno_strains]
tree = keep.tip(full_tree, tip = good_tips)
write.tree(tree, file = 'pheno_tree.nwk')
```

Now for pyseer 
```{bash}
#optional chunk, I don't include python environment in writeup 
#source activate python
#python phylogeny_distance.py --lmm pheno_tree.nwk > phylogeny_K.tsv
```

There's our kinship matrix. We start with a phylogeny-based K in this case because previous work (Karasov et al. 2018) made this phylogeny after removing recombination tracts within OTU5. This estimates true lineages within OTU5, and should provide more accurate distance measurements. 



Now for GWAS! We use a linear mixed model, which gives us heritability estimates for each gene, as well as estimating the polygenic score as a random effect from K. See model (adapted from EMMA: Kang et al., n.d.) below: $$y = βX + K \textbf{u} + \textbf{e}$$ where y is the phenotype vector, beta is the coefficients vector (fixed effect of every variant), K is our Kinship matrix, $\textbf{u}$ is the random polygenic effect, and $\textbf{e}$ is the residuals. We'll implement in pyseer:
```{bash}
#I wound't run this chunk - you can install pyseer and run if you wish, but this is a pretty
#intensive process - takes a while on the cluster

#pyseer --phenotypes phen_prelim_Oct2019.tsv --pres pa_data.tsv --similarity phylogeny_K.tsv --lmm > pyseer_PA.assoc
```

Let's make a function for our analysis (that we can go back to)

```{R}
make_qq <- function(path){

  PA_assoc <- read_tsv(path)
  
  #qqplot
  pvals <- dplyr::select(PA_assoc, `lrt-pvalue`) %>% rename(assoc_ps = `lrt-pvalue`)
  rand_ps <- runif(nrow(pvals))
  pvals <- add_column(pvals, rand_ps) %>% 
    transmute(`Theoretical Quantiles` = -log10(sort(rand_ps)), 
              `Sample Quantiles` = -log10(sort(assoc_ps)))
  
  ggplot(data = pvals) + 
    aes(x = `Theoretical Quantiles`, y = `Sample Quantiles`) + 
    geom_point(alpha = 0.3) + 
    geom_abline(slope = 1, color = 'red') + 
    labs(title = "QQ Plot of PA Data") + 
    theme_classic() 
}

make_manhattan <- function(path){
  PA_assoc <- read_tsv(path)
  #manhattan plot
  gwas_data <- PA_assoc %>% mutate(log_p = -1 * log10(`lrt-pvalue`))
  ggplot(data = gwas_data, aes(x = variant, y = log_p)) + 
    geom_point() + 
    geom_abline(aes(slope = 0, intercept = -1 * log10(0.05 / nrow(gwas_data))), color = 'red') + 
    labs(x = 'gene number', y = '-log_10_p', title = "Manhattan Plot of PA Data") + 
    theme_classic()
}

make_qq('pyseer_PA.assoc')
make_manhattan('pyseer_PA.assoc')

```


With this initial GWAS, we see some issues. Our p values are not well-controlled near p = 0.1 to 0.01 (theoretical quantiles 1 and 2). This means we probably have some issues with population structure (unlikely for tens of thousands of variants to be causally associated). In addition, the 'shelf' of p values we can see both on the Manhattan plot and on the qq plot confirms this: these are several genes in LD, being carried at the same frequency and in a lineage with a truly casual gene.


To remedy this, we can have a more stringent minor allele frequency (MAF) to only include variants at a higher percentage present in the strains (more power to detect associations, from less multiple testing, and more insertion/recombination with these alleles, at the cost of losing some variants). We can also check our data to see if it's normal - we want to see a normally distributed random effect. 

Let's try with MAF >= 0.05:
```{bash}
#pyseer --phenotypes phen_prelim_Oct2019.tsv --pres pa_data.tsv --similarity phylogeny_K.tsv --lmm --min-af 0.05 --max-af 0.95 > pyseer_PA_AFcutoffs.assoc
```

```{R}
make_qq('pyseer_PA_AFcutoffs.assoc')
make_manhattan('pyseer_PA_AFcutoffs.assoc')
```
We only test 2238 variants, but this looks a lot more well-controlled in the Manhattan plot. In the qq plot, we still see a slight inflation of p values, but this looks better. Let's save these significant hits.
```{r}
gwas_data <- read_tsv('pyseer_PA_AFcutoffs.assoc') 
sig_genes_afcutoffs <- gwas_data %>% filter(`lrt-pvalue` < (0.05 / nrow(gwas_data)))
```

Let's look at our phenotype.
```{r}
phen_prelim <- read_tsv('phen_prelim_Oct2019.tsv')
phen <- phen_prelim$blup


#boxcox 
phen_all_pos <- phen -  min(phen) + 0.0001
boxcox(phen_all_pos ~ 1)
#lambda = 1.5
phen_box <- phen_all_pos ^ 1.5

#exponentiated
phen_exp <- 10^phen

par(mfrow = c(1,3))
hist(phen); hist(phen_exp); hist(phen_box)
qqPlot(phen); qqPlot(phen_exp); qqPlot(phen_box)
```

The Box-Cox with $\lambda = 1.5$ appears to be the most normally distributed. Let's try a GWAS with this. 


```{r}
normal_phen <- phen_prelim %>% add_column(phen_box) %>% transmute(sample, blups = phen_box)
write_tsv(normal_phen, 'phen_prelim_Oct2019_normalized.tsv')
```
```{bash}
#pyseer --phenotypes phen_prelim_Oct2019_normalized.tsv --pres pheno_pa_data.tsv --similarity phylogeny_K.tsv --lmm > pyseer_PA_normalized.assoc
```
```{r}
make_qq('pyseer_PA_normalized.assoc')
make_manhattan('pyseer_PA_normalized.assoc')
```
We still seem to have plenty of issues with population structure in this graph. Let's check normalized data with allele frequency cutoffs. 

```{bash}
#pyseer --phenotypes phen_prelim_Oct2019_normalized.tsv --pres pheno_pa_data.tsv --similarity phylogeny_K.tsv --lmm --min-af 0.05 --max-af 0.95 > pyseer_PA_normalized_AFcutoffs.assoc
```
```{r}
make_qq('pyseer_PA_normalized_AFcutoffs.assoc')
make_manhattan('pyseer_PA_normalized_AFcutoffs.assoc')
```

This appears to be the best result so far. We only have two significant alleles, but our p values are well controlled well into 0.01, and there still appear to be several associated genes. Let's save these genes.
```{r}
gwas_data <- read_tsv('pyseer_PA_normalized_AFcutoffs.assoc') 
sig_genes_normalized <- gwas_data %>% filter(`lrt-pvalue` < (0.05 / nrow(gwas_data)))

sig_genes_normalized
sig_genes_afcutoffs
```
Two of them are the same: 5463 and 5466, and all four have decent heritability estimates. Let's take a look at the two robust ones

```{#python}
#do not run this chunk. It relies on a pickled library of several GB 

#from get_cluster_data import load_sorted_clusters
#sorted_clusters = load_sorted_clusters('all_clusters_postprocessed.cpk')
#sorted_clusters[5462, 5465]
```
(I'm not running the chunk above right now, but in general, we'd find the following genes: Phage Holin Family (Lysis Protein S) and Hypothetical Protein (both with unique gene IDs, etc.)



References:


Corbett, E. L., Watt, C. J., Walker, N., Maher, D., Williams, B. G., Raviglione, M. C., & Dye, C. (2003, May 12). The growing burden of tuberculosis: Global trends and interactions with the HIV epidemic. Archives of Internal Medicine, Vol. 163, pp. 1009–1021. https://doi.org/10.1001/archinte.163.9.1009

Ding, W., Baumdicker, F., & Neher, R. A. (2018). panX: pan-genome analysis and exploration. Nucleic Acids Research, 46(1), e5. https://doi.org/10.1093/nar/gkx977

Holley, G., & Melsted, P. (2020). Bifrost: Highly parallel construction and indexing of colored and compacted de Bruijn graphs. Genome Biology, 21(1), 249. https://doi.org/10.1186/s13059-020-02135-8
Kang, H. M., Zaitlen, N. A., Wade, C. M., Kirby, A., Heckerman, D., Daly, M. J., & Eskin, E. (n.d.). Efficient Control of Population Structure in Model Organism Association Mapping. https://doi.org/10.1534/genetics.107.080101

Karasov, T. L., Almario, J., Friedemann, C., Ding, W., Giolai, M., Heavens, D., … Weigel, D. (2018). Arabidopsis thaliana and Pseudomonas Pathogens Exhibit Stable Associations over Evolutionary Timescales. Cell Host and Microbe, 24(1), 168-179.e4. https://doi.org/10.1016/j.chom.2018.06.011

Lees, J. A., Galardini, M., Bentley, S. D., Weiser, J. N., Corander, J., & Stegle, O. (n.d.). pyseer: a comprehensive tool for microbial pangenome-wide association studies. https://doi.org/10.1093/bioinformatics/bty539












